ipfscluster_test.go

```diff
@@ -816,6 +816,275 @@ func TestClustersReplication(t *testing.T) {
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err == nil {
+		t.Error("Pin should have failed as rplMin cannot be satisfied")
+	}
+	t.Log(err)
+	if !strings.Contains(err.Error(), fmt.Sprintf("not enough peers to allocate CID")) {
+		t.Fatal(err)
+	}
+}
+
+// This tests checks that repinning something that has becomed
+// underpinned actually changes nothing if it's sufficiently pinned
+func TestClustersReplicationMinMaxNoRealloc(t *testing.T) {
+	if nClusters < 5 {
+		t.Skip("Need at least 5 peers")
+	}
+
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = 1
+		c.config.ReplicationFactorMax = nClusters
+	}
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Shutdown two peers
+	clusters[nClusters-1].Shutdown()
+	clusters[nClusters-2].Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	err = clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err := clusters[0].PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(p.Allocations) != nClusters {
+		t.Error("allocations should still be nCluster even if not all available")
+	}
+
+	if p.ReplicationFactorMax != nClusters {
+		t.Error("rplMax should have not changed")
+	}
+}
+
+// This test checks tat repinning something that has becomed
+// underpinned does re-allocations when it's not sufficiently
+// pinned anymore
+func TestClustersReplicationMinMaxRealloc(t *testing.T) {
+	if nClusters < 5 {
+		t.Skip("Need at least 5 peers")
+	}
+
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = 3
+		c.config.ReplicationFactorMax = 4
+	}
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	delay()
+
+	p, err := clusters[0].PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	firstAllocations := p.Allocations
+
+	peerIDMap := make(map[peer.ID]*Cluster)
+	for _, a := range clusters {
+		peerIDMap[a.id] = a
+	}
+
+	// kill two of the allocations
+	alloc1 := peerIDMap[firstAllocations[0]]
+	alloc2 := peerIDMap[firstAllocations[1]]
+	safePeer := peerIDMap[firstAllocations[2]]
+
+	alloc1.Shutdown()
+	alloc2.Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	// Repin - (although this might have been taken of if there was an alert
+	err = safePeer.Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err = safePeer.PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	secondAllocations := p.Allocations
+
+	if len(secondAllocations) != minInt(nClusters-2, 4) {
```

Comment:
We should have a more precise error message here and ideally remove the redundancy with the conditional on 1083 which will never get hit because of this conditional.  Maybe put this one after 1083 and report "insufficient reallocation, could have allocated to %d nodes but instead only allocated to %d nodes".

Reply:
done, but not that t.Error() does not abort the tests, it just marks as failure, so the second conditional would have happened too (unlike with t.Fatal).

Reply:
ah thanks for the clarification